homeassistant/components/climate/hive.py

```diff
@@ -0,0 +1,204 @@
+"""
+Support for the Hive devices.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/hive/
+"""
+import logging
+from datetime import datetime
+
+from homeassistant.components.climate import ENTITY_ID_FORMAT, ClimateDevice
+from homeassistant.const import ATTR_TEMPERATURE, TEMP_CELSIUS
+
+DEPENDENCIES = ['hive']
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_platform(hass, config, add_devices, hivedevice, discovery_info=None):
+    """Set up Hive climate devices."""
+    session = hass.data.get('DATA_HIVE')
+
+    add_devices([HiveClimateEntity(hass, session, hivedevice)])
+
+
+class HiveClimateEntity(ClimateDevice):
+    """Hive Climate Device."""
+
+    def __init__(self, hass, Session, HiveDevice):
+        """Initialize the Climate device."""
+        self.node_id = HiveDevice["Hive_NodeID"]
+        self.node_name = HiveDevice["Hive_NodeName"]
+        self.device_type = HiveDevice["HA_DeviceType"]
+        self.hass = hass
+        self.session = Session
+
+        if self.device_type == "Heating":
+            set_entity_id = "Hive_Heating"
+            self.session.heating = self.session.core.Heating()
+        elif self.device_type == "HotWater":
+            set_entity_id = "Hive_HotWater"
+            self.session.hotwater = self.session.core.Hotwater()
+        if self.node_name is not None:
+            set_entity_id = set_entity_id + "_" \
+                            + self.node_name.replace(" ", "_")
+        self.entity_id = ENTITY_ID_FORMAT.format(set_entity_id.lower())
+
+        self.hass.bus.listen('Event_Hive_NewNodeData', self.handle_event)

```

Comment:
I'm not sure I understand what you're doing with this event. You update all the devices when something changes? How about maintaining a list of all Hive entities in `hass.data[DOMAIN]["devices"]["climate"]` or something, and then scheduling an update for all of them? Since this event is used only within this component, it probably is not of interest outside of it? This would reduce the number of events flying across the event bus.

Reply:
The structure of the Hive Component and Platforms is that there is one instance of the library which is in ```hass.data["HIVE_DATA"]``` and this is shared between all of the Hive Platforms.
If any of the entities make a change, such as turn on a switch, change the brightness of a light or set a new temperature for example, this will cause the library to fetch the latest set of data for all of the Hive entities. When this is done an event is fired to trigger all the devices to fetch that new data which is now stored locally in the instance in ```hass.data["HIVE_DATA"]```. So when the events are fires there is not any new api call made to Hive, all the traffic is local and retrieving new data from the local cache.
Sorry for the wall of text, but hopefully it helps clarify what is going on?

Reply:
I think I understand. However, there are two things:
* If I switch on a switch via home assistant, why does this have to cause the state of for example all lights to be updated? How are these two things related?
* If you want to say "update all the entities", why do you use an event for that? I suggest this: Have some central list of all hive entities. Either inside the component object (i.e., every entity needs to call `hass.data["HIVE_DATA"].register_entity(…)` or something in its constructor) or directly attached to the hass.data["HIVE_DATA"] object. Then, if a "round of updates" is to be performed, whoever wants to do that calls `hass.data["HIVE_DATA"].update_all()`, which in turn does something like 
 ```
for entity in self.all_entities:
  entity.update()
 ```

Reply:
so whenever I need to get some entities to update I would need to run through the same code in each platform file in each function which needs to make that trigger to get the entities to update?

what does the ```...``` in ```hass.data["HIVE_DATA"].register_entity(…)``` represent?

To explain further.
The library only allows api calls to Hive at most once a minute, by default once every two minutes.
If you have a sensor just trigger or a switch just turned on (not via HA) just moments after the last api call then you have to wait the full time to get that new state

during an api call to the Hive api, the Hive api always returns the full device current status and details each time you change something, so you turn on a light, you get the full device current status.
The reason I put the code in there to update all entities was so that each time there is new data from the external web hosted api retrieved and added to the local to HA cache of data then there is fresh data that all the devices will benefit from in updating to. so the all do an update to get the latest data from the local cache. so if a sensor was triggered or the heating state changed, turning the light on would get that latest data in to the entities.

using the event to trigger that was the only way I could find in the online HA documentation to do that.
If I can achieve the same thing by some better method then I am all up for that. and if there is a better way at controlling which entities get updated, even better
I will just need a hand in moving to that new approved / allowed method

Reply:
Okay, if the API call to set the switch returns all status information, it makes sense to then update it all.

Regarding the event thing: I'm not saying that each entity should hit the API. I'm saying that triggering the update should happen by calling a function, not by firing an event. Currently you fire an event, and that event causes all entities to call `handle_event`, right? Say you had a list of all entities. Then you could just iterate that list and call `handle_event` on each of the list's items, i.e., on each entity. That would then result in exactly the same outcome as firing the event, am I correct?

That's what I'm suggesting. Maintain a list of all hive entities somewhere (in hass.data, or at the main component object) and instead of firing the event, iterate that list and call `handle_event` (which should then be renamed) on all of them. That gets rid of the event which is currently just used for inner-platform communication.

Reply:
Sounds good ;)

Do you have any code snippets on how I could for example in ```light``` ```turn_on``` cycle round all the entities to trigger them to then call ```handle_event```, including those non ```light``` entities?

I guess there will need to be a ```handle_event``` (renamed) in each platform file?

Reply:
I don't have a code snippet ready, sorry. Here's what I would do:

* Have (as you said correctly) a `handle_event()` (perhaps `handle_update()`?) method in each platform
* Add a list of all entities to the component: in the constructor, do `self._entities = []`
* Add a `register_entity(self, entity)` method in the component, that just appends the passed entity into the `self._entities` list
* Make every platform call `hass.data[DOMAIN].register_entity(self)`, i.e., add the entity to the list of all entities
* Add an `update_all(self)` method to the component, which does something along the lines of:
```
for entity in self._entites:
  entity.handle_update()
```
* Where you fired the event so far, call `hass.data[DOMAIN].update_all()`

Reply:
Sorry, I think this is beyond me.
I thought it was prohibited to add code to the Component so that everything could call it?
Is there a component with platforms that does this already that I could take a look at?

do I really have to remove the current method used? I really don't know how to proceed with this any other way.

would ```self.hass.data[DOMAIN].entitylist.append(entity)``` work in gathering the full list of entities and make available to all platforms?

Reply:
Huh, platforms may not call methods on the component? That's new to me (which doesn't mean it's not true. :-D) Where did you find that?

But now that I dig deeper into this, the dispatcher seems to be built exactly for this job. Have a look at https://home-assistant.io/developers/code_review_component/ , last point. You can find quite a few examples in the code base, just search for `dispatcher_connect`. Basically, `dispatcher_connect` replaces the "registering" call from platform to component. Then, `dispatcher_send` replaces the call from component to platform. Looks similar to an event, but it's not actually an event on the event bus, and works pretty much exactly how I've outlined above. ;)

Reply:
:)
some things fabaff asked me to change during his review of this.

I saw that note on the page you linked, but I could not find any documentation of anything existing already which used it.
do you have a name of platform and component which uses this that I could take a look at?

Reply:
Hi @tinloaf 

How is it now, will the latest code be ok?
I have used what I think you outlined before, but not used any Component calls.

If this new updating method is ok, does this mean that this PR is good to go?